### PR TITLE
Add Counterfactual Safes retrieval/creation endpoints

### PR DESCRIPTION
**Pull request by hectorgomezv**
## Summary
This PR adds a subset of the Counterfactual Safes storage controller endpoints. Specifically:
```
GET /v1/accounts/:address/storage/counterfactual-safes/:chainId/:predictedAddress
PUT /v1/accounts/:address/storage/counterfactual-safes
```

The first one retrieves an existing Counterfactual Safe by the creator's account address, the chain ID, and the predicted address of the Safe to be deployed.

The second one checks if a Counterfactual Safes already exists for the combination of the creator's account address, chain ID, and predicted address. If such a Counterfactual Safe exists, it's returned. Otherwise, it gets created. This endpoint expects the following payload as request body:

```
  chainId: string;
  fallbackHandler: `0x${string}`;
  owners: `0x${string}`[];
  predictedAddress: `0x${string}`;
  saltNonce: string;
  singletonAddress: `0x${string}`;
  threshold: number;
```

**Note: Endpoints for Counterfactual Safes deletion will also be added in a separate PR.**

## Changes
- Add `GET /v1/accounts/:address/storage/counterfactual-safes/:chainId/:predictedAddress`
- Add `PUT /v1/accounts/:address/storage/counterfactual-safes`

### CODE REVIEW

File: src/domain/interfaces/counterfactual-safes.datasource.interface.ts [hunk 1]
```diff
@@ -13,6 +13,7 @@ export interface ICounterfactualSafesDatasource {
   }): Promise<CounterfactualSafe>;
 
   getCounterfactualSafe(args: {
+    account: Account;
```
Comment (hectorgomezv) [inline src/domain/interfaces/counterfactual-safes.datasource.interface.ts:16]: I've added the `account`  here, as it's always available on the domain level, and it narrows down the query surface, improving performance.

File: src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    );
+    if (!counterfactualSafeSetting?.enabled) {
+      this.loggingService.warn({
+        message: `Account ${args.address} does not have CounterfactualSafes enabled`,
+      });
+      throw new GoneException();
+    }
+  }
+
+  private async checkCounterfactualSafeDataTypeIsActive(): Promise<AccountDataType> {
```
Comment (hectorgomezv) [inline src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts:125]: The result of this function shouldn't change until a new migration is applied. `this.accountsRepository.getDataTypes()` already involves permanent caching. I think that should be enough,  but this function could be memoized eventually.

File: src/domain/accounts/counterfactual-safes/entities/__tests__/counterfactual-safe.builder.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { IBuilder, Builder } from '@/__tests__/builder';
+import { CounterfactualSafe } from '@/domain/accounts/counterfactual-safes/entities/counterfactual-safe.entity';
+import { faker } from '@faker-js/faker';
+import { getAddress } from 'viem';
+
+export function counterfactualSafeBuilder(): IBuilder<CounterfactualSafe> {
```
Comment (hectorgomezv) [inline src/domain/accounts/counterfactual-safes/entities/__tests__/counterfactual-safe.builder.ts:6]: The domain entity `CounterfactualSafe` still uses snake_case syntax, but a forthcoming PR will restrict the usage of snake_case to the datasources layer, as discussed in https://github.com/safe-global/safe-client-gateway/pull/1773#discussion_r1684190135

File: src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    }
+    const [account] = await Promise.all([
+      this.accountsRepository.getAccount({
+        authPayload: args.authPayload,
+        address: args.address,
+      }),
+      this.checkCounterfactualSafesIsEnabled({
+        authPayload: args.authPayload,
+        address: args.address,
+      }),
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts]: Changed in [b853932](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/b8539322780b888222d7944c6315c37ffc71adfb)

File: src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts]: Changed in [b853932](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/b8539322780b888222d7944c6315c37ffc71adfb)

File: src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  /**
+   * Gets or creates a Counterfactual Safe.
+   * Checks that the account has the CounterfactualSafes data setting enabled.
+   *
+   * If the Counterfactual Safe exists, it returns it.
+   * If the Counterfactual Safe does not exist, it's created.
+   */
+  async getOrCreateCounterfactualSafe(args: {
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts]: Changed in [b853932](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/b8539322780b888222d7944c6315c37ffc71adfb)

File: src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        message: `Account ${args.address} does not have CounterfactualSafes enabled`,
+      });
+      throw new GoneException();
+    }
+  }
+
+  private async checkCounterfactualSafeDataTypeIsActive(): Promise<AccountDataType> {
+    const dataTypes = await this.accountsRepository.getDataTypes();
+    const counterfactualSafeDataType = dataTypes.find(
+      (dataType) => dataType.name === 'CounterfactualSafes',
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts]: Changed in [b853932](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/b8539322780b888222d7944c6315c37ffc71adfb)

File: src/routes/accounts/counterfactual-safes/entities/counterfactual-safe.entity.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  @ApiProperty()
+  predictedAddress: `0x${string}`;
+  @ApiProperty()
+  saltNonce: string;
+  @ApiProperty()
+  singletonAddress: `0x${string}`;
+  @ApiProperty()
+  threshold: number;
+  @ApiProperty()
+  accountId: string;
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/routes/accounts/counterfactual-safes/entities/counterfactual-safe.entity.ts]: Changed in [b853932](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/b8539322780b888222d7944c6315c37ffc71adfb)

File: src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts [hunk 9]
```diff
@@ -0,0 +1,587 @@
...
+          `/v1/accounts/${address}/storage/counterfactual-safes/${chain.chainId}/${counterfactualSafe.predicted_address}`,
+        )
+        .set('Cookie', [`access_token=${accessToken}`])
+        .expect(403);
+
+      expect(accountsRepository.getAccount).not.toHaveBeenCalled();
+      expect(
+        counterfactualSafesDataSource.getCounterfactualSafe,
+      ).not.toHaveBeenCalled();
+    });
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts]: Changed in [b853932](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/b8539322780b888222d7944c6315c37ffc71adfb)

File: src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts [hunk 10]
```diff
@@ -0,0 +1,587 @@
...
+        .expect(403);
+
+      expect(accountsRepository.getAccount).not.toHaveBeenCalled();
+      expect(
+        counterfactualSafesDataSource.getCounterfactualSafe,
+      ).not.toHaveBeenCalled();
+      expect(
+        counterfactualSafesDataSource.createCounterfactualSafe,
+      ).not.toHaveBeenCalled();
+    });
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts]: Changed in [b853932](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/b8539322780b888222d7944c6315c37ffc71adfb)

File: src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts [hunk 11]
```diff
@@ -0,0 +1,587 @@
...
+        .expect(403);
+
+      expect(accountsRepository.getAccount).not.toHaveBeenCalled();
+      expect(
+        counterfactualSafesDataSource.getCounterfactualSafe,
+      ).not.toHaveBeenCalled();
+    });
+
+    it('returns 403 if signer_address is not a valid Ethereum address', async () => {
+      const address = faker.string.hexadecimal() as `0x${string}`;
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts]: Changed in [b853932](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/b8539322780b888222d7944c6315c37ffc71adfb)

File: src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts [hunk 12]
```diff
@@ -0,0 +1,587 @@
...
+      expect(
+        counterfactualSafesDataSource.getCounterfactualSafe,
+      ).not.toHaveBeenCalled();
+      expect(
+        counterfactualSafesDataSource.createCounterfactualSafe,
+      ).not.toHaveBeenCalled();
+    });
+
+    it('returns 403 if signer_address is not a valid Ethereum address', async () => {
+      const address = faker.string.hexadecimal() as `0x${string}`;
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts]: Changed in [b853932](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/b8539322780b888222d7944c6315c37ffc71adfb)

File: src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+import { Body, Controller, Get, Param, Put, UseGuards } from '@nestjs/common';
+import { ApiOkResponse, ApiTags } from '@nestjs/swagger';
+
+@ApiTags('accounts')
+@Controller({ path: 'accounts', version: '1' })
+export class CounterfactualSafesController {
+  constructor(private readonly service: CounterfactualSafesService) {}
+
+  @ApiOkResponse({ type: CounterfactualSafe })
+  @Get(':address/storage/counterfactual-safes/:chainId/:predictedAddress')
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.ts]: Agreed. Changed in [121f316](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/121f316de3e021ca30365e34f4e62d7abf4890dc)

**Comment by hectorgomezv** [comment]
I've made two additional changes to the PR (cc. @iamacook):

- I've included the DTO route entity (it wasn't properly shown on Swagger)
- I've renamed `ICounterfactualSafesRepository.upsertCounterfactualSafe` to `ICounterfactualSafesRepository.createCounterfactualSafe`, as I think it makes more sense: no "upsert" is made as the Counterfactual Safe are immutable once they are created, so no "update" is possible. Please let me know if you think we could name this function differently.

File: src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts [hunk 14]
```diff
@@ -0,0 +1,587 @@
...
+        .set('Cookie', [`access_token=${accessToken}`])
+        .expect(403);
+
+      expect(accountsRepository.getAccount).not.toHaveBeenCalled();
+      expect(
+        counterfactualSafesDataSource.getCounterfactualSafe,
+      ).not.toHaveBeenCalled();
+    });
+
+    it('returns 403 if chain_id is not a valid chain ID', async () => {
```
Comment (iamacook) [inline src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts:255]: What do you think about adding another test case that ensure the token is valid if signed on a different chain?

**Review by iamacook — APPROVED** [review]

File: src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts [hunk 15]
```diff
@@ -138,6 +138,52 @@ describe('CounterfactualSafesController', () => {
...
+          .with('enabled', true)
+          .build(),
+      ]);
+      counterfactualSafesDataSource.getCounterfactualSafe.mockResolvedValue(
+        counterfactualSafe,
+      );
+
+      await request(app.getHttpServer())
+        .get(
+          `/v1/accounts/${address}/counterfactual-safes/${faker.string.numeric()}/${counterfactualSafe.predicted_address}`,
```
Comment (iamacook) [inline src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts:171]:
```suggestion
          `/v1/accounts/${address}/counterfactual-safes/${faker.string.numeric({ exclude: authPayloadDto.chain_id })}/${counterfactualSafe.predicted_address}`,
```

**Review by iamacook — APPROVED** [review]

File: src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts (hunk 14 above)
Comment (hectorgomezv) [reply src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts:301]: Added in [b79e675](https://github.com/safe-global/safe-client-gateway/pull/1789/commits/b79e67575b8aeabec21488aa9cff2487874946ab)

Additional thought: I feel like we are duplicating a lot of checks on these "403 assertion tests". Essentially we are testing `JwtService` using each controller as a proxy. I wouldn't tackle it on this PR, but I'd remove all these redundant tests. `JwtService` tests already test that class in isolation.

In the controllers, we would just assert they call `JwtService` (where applicable). Do you think it is worth repeating ourselves on each endpoint test for some reason?

Comment (iamacook) [reply src/routes/accounts/counterfactual-safes/counterfactual-safes.controller.spec.ts:301]: Makes sense to me!

**Pull request merged**